pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+elementary gates.
+
+Qubit architectures
+-------------------
+
+.. autosummary::
+
+    BasisStatePreparation
+    MottonenStatePreparation
+
+Code details
+^^^^^^^^^^^^
+"""
+import math
+from collections.abc import Iterable
+from scipy import sparse
+
+import numpy as np
+import pennylane as qml
+
+# pylint: disable=len-as-condition,arguments-out-of-order
+
+
+def gray_code(rank):
+    """Generates the Gray code of given rank.
+
+    Args:
+        rank (int): rank of the Gray code (i.e. number of bits)
+    """
+    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]
+
+
+def BasisStatePreparation(basis_state, wires):
+    r"""
+    Prepares a basis state on the given wires using a sequence of Pauli X gates.
+
+    Args:
+        basis_state (array): Input array of shape ``(N,)``, where N is the number of wires
+            the state preparation acts on. ``N`` must be smaller or equal to the total
+            number of wires.
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+
+    if not isinstance(wires, Iterable):
+        raise ValueError("Wires must be passed as a list of integers; got {}.".format(wires))
+
+    if not len(basis_state) == len(wires):
+        raise ValueError(
+            "Number of qubits must be equal to the number of wires, which is {}; "
+            "got {}.".format(len(wires), len(basis_state))
+        )
+
+    if any([x not in [0, 1] for x in basis_state]):
+        raise ValueError("Basis state must only consist of 0s and 1s, got {}".format(basis_state))
+
+    for wire, state in zip(wires, basis_state):
+        if state == 1:
+            qml.PauliX(wire)
+
+
+def _matrix_M_entry(row, col):
+    """Returns one entry for the matrix that maps alpha to theta.
+
+    Args:
+        row (int): one-based row number
+        col (int): one-based column number
+
+    Returns:
+        (float): transformation matrix entry at given row and column
+    """
+    # (col >> 1) ^ col is the Gray code of col
+    b_and_g = row & ((col >> 1) ^ col)
+    sum_of_ones = 0
+    while b_and_g > 0:
+        if b_and_g & 0b1:
+            sum_of_ones += 1
+
+        b_and_g = b_and_g >> 1
+
+    return (-1) ** sum_of_ones
+
+
+def _compute_theta(alpha):
+    """Calculates the rotation angles from the alpha vector.
+
+    Args:
+        alpha (array[float]): alpha parameters
+
+    Returns:
+        (array[float]): rotation angles theta
+    """
+    k = np.log2(alpha.shape[0])
+    factor = 2 ** (-k)
+
+    theta = sparse.dok_matrix(alpha.shape, dtype=np.float64)  # type: sparse.dok_matrix
+
+    for row in range(alpha.shape[0]):
+        # Use transpose of M:
+        entry = sum([_matrix_M_entry(col, row) * a for (col, _), a in alpha.items()])
+        entry *= factor
+        if abs(entry) > 1e-6:
+            theta[row, 0] = entry
+
+    return theta
+
+
+def _uniform_rotation_dg(gate, alpha, control_wires, target_wire):
```

Comment:
Writing out the word `dagger` in the function name could increase readability of the code (for each of the later functions as well).